pep-0xxx.rst

```diff
@@ -0,0 +1,1250 @@
+
+            The buffer objects must be either file objects or objects that
+            implement the buffer protocol.
+            """
+
+
+Socket
+~~~~~~
+
+The socket-wrapper ABC will be defined by the ``TLSWrappedSocket`` ABC, which
+has the following definition::
+
+    class TLSWrappedSocket(metaclass=ABCMeta):
+        # The various socket methods all must be implemented. Their definitions
+        # have been elided from this class defintion in the PEP because they
+        # aren't instructive.
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[CipherSuite]:
+            """
+            Returns the CipherSuite entry for the cipher that has been
+            negotiated on the connection. If no connection has been negotiated,
+            returns ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[Union[NextProtocol, bytes]]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If the negotiated protocol is one of the protocols defined in the
+            ``NextProtocol`` enum, the value from that enum will be returned.
+            Otherwise, the raw bytestring of the negotiated protocol will be
+            returned.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def unwrap(self) -> socket.socket:
+            """
+            Cleanly terminate the TLS connection on this wrapped socket. Once
+            called, this ``TLSWrappedSocket`` can no longer be used to transmit
+            data. Returns the socket that was wrapped with TLS.
+            """
+
+
+Buffer
+~~~~~~
+
+The buffer-wrapper ABC will be defined by the ``TLSWrappedBuffer`` ABC, which
+has the following definition::
+
+    class TLSWrappedBuffer(metaclass=ABCMeta):
+        @abstractmethod
+        def read(self, amt=None: int) -> bytes:
+            """
+            Read up to ``amt`` bytes of data from the input buffer and return
```

Comment:
I don't know about *reasonable*, but it's extremely common in higher-level Python programs. For example, many HTTP libraries are implemented in exactly this way.

*However*, I agree that it's not a reasonable idiom for unbuffered objects like this one. Generally `read()` as "read to EOF" is only a sensible idiom when `read()` is guaranteed to return `amt`.

As for `suppress_ragged_eofs`, that's a trickier issue. @njsmith has brought this up once before and I've not taken an enormously forceful position on it, but *generally speaking* `suppress_ragged_eofs` is a bad option for security but a good one for cleanliness in certain protocols.

In this instance, it may be worth expressing at length the behaviour here. The expected behaviour is that `read()` (and friends) will stick to the Python behaviour of always returning `b''` once EOF is reached. *However*, we may elect to say that "ragged EOFs" (which is a TCP FIN or RST without a preceding TLS CloseNotify) *MAY* cause an exception to be raised (which we'd want to define in this ABC) instead of an EOF condition. That would require that callers expect and handle that exception, which is a good idea anyway. (HTTP/1.1 and HTTP/2 implementations will almost always just swallow it, as these protocols are not vulnerable to length-truncation attacks: other protocols may want to expose those exceptions to the user.)